### PR TITLE
feat: implement retry policy for jobs

### LINKED ISSUE #40: Implement retry and backoff behavior
## Summary
Implement: retry and backoff behavior.

## Scope
- deliver the code, runtime wiring, and directly-coupled tests for this capability
- cover the happy path, important failure paths, and boundary conditions that are part of the feature itself
- update only the documentation that is tightly coupled to the implementation

## Not In Scope
- broad docs sweeps, cross-milestone cleanup, or unrelated platform work

## Acceptance Criteria
- the capability works end-to-end for the supported scenarios in this milestone
- important error and recovery paths are covered by tests
- public behavior, caveats, and unsupported cases are clear from the resulting implementation

### PR DESCRIPTION
**Pull request by niteshpurohit**
- Introduced a new `Karya::RetryPolicy` class to manage retry and backoff behavior for jobs.
- Added configuration options for maximum attempts, base delay, multiplier, and optional maximum delay.
- Integrated retry policy into the worker configuration, allowing per-job overrides.
- Updated job attributes to include retry policy and next retry timestamp.
- Enhanced worker behavior to handle retry logic, transitioning jobs to a `retry_pending` state when necessary.
- Implemented tests to validate retry policy functionality and integration with job processing.
- Updated documentation to reflect new retry policy features and usage.

closes: #40

### CODE REVIEW

File: core/karya/lib/karya/queue_store/in_memory/store_state.rb [hunk 1]
```diff
@@ -41,6 +43,14 @@ def delete_queue(queue)
           queued_job_ids_by_queue.delete(queue)
         end
 
+        def register_retry_pending(job_id)
+          retry_pending_job_ids << job_id unless retry_pending_job_ids.include?(job_id)
```
[the remark this answers is not in the document]
Comment (niteshpurohit) [reply core/karya/lib/karya/queue_store/in_memory/store_state.rb]: Fixed. `StoreState#register_retry_pending` now always returns `retry_pending_job_ids`, and the Ruby implementation matches the RBS return type.

[the remark this answers is not in the document]
Comment (niteshpurohit) [reply core/karya/lib/karya/queue_store/in_memory/store_state.rb]: Addressed. Kept the ordered array for promotion scans, but added a parallel membership index so duplicate checks are now O(1) instead of `Array#include?`.

File: core/karya/sig/karya/queue_store/in_memory.rbs [hunk 2]
```diff
@@ -111,6 +113,10 @@ module Karya
 
         def delete_queue: (String queue) -> Array[String]?
 
+        def register_retry_pending: (String job_id) -> Array[String]
```
[the remark this answers is not in the document]
Comment (niteshpurohit) [reply core/karya/sig/karya/queue_store/in_memory.rbs:117]: Fixed from the Ruby side. `register_retry_pending` now consistently returns `Array[String]`, so the existing RBS return type is accurate.

File: core/karya/lib/karya/worker.rb [hunk 3]
```diff
@@ -81,6 +81,10 @@ def lifecycle
       configuration.lifecycle
     end
 
+    def retry_policy
+      configuration.retry_policy
+    end
```
[the remark this answers is not in the document]
Comment (niteshpurohit) [reply core/karya/lib/karya/worker.rb:87]: Fixed. Added `require_relative 'retry_policy'` to `karya/worker.rb` so direct `require 'karya/worker'` loads the retry policy constant needed by worker configuration.

File: core/karya/sig/karya/queue_store/in_memory.rbs [hunk 4]
```diff
@@ -83,6 +83,7 @@ module Karya
         @rate_limit_admissions_by_key: Hash[String, Array[Time]]
 
         @queued_job_ids_by_queue: Hash[String, Array[String]]
+        @retry_pending_job_ids: Array[String]
```
[the remark this answers is not in the document]
Comment (niteshpurohit) [reply core/karya/sig/karya/queue_store/in_memory.rbs:86]: Fixed. Added `@retry_pending_job_ids_index: Hash[String, bool]` to `StoreState` in the RBS so the signature matches the Ruby implementation.

File: core/karya/spec/karya/worker_integration_spec.rb [hunk 5]
```diff
@@ -81,6 +82,74 @@ def stored_job(job_id)
...
+      retry_policy: retry_policy,
+      clock: -> { retrying_clock.next }
+    )
+
+    first_result = worker.work_once
+    second_result = worker.work_once
+
+    expect(first_result.state).to eq(:retry_pending)
+    expect(first_result.next_retry_at).to eq(base_time + 8)
+    expect(stored_job('job-retry').attempt).to eq(2)
```
[the remark this answers is not in the document]
Comment (niteshpurohit) [reply core/karya/spec/karya/worker_integration_spec.rb]: Fixed. Updated the integration spec to assert `first_result.attempt == 1` while the job is `retry_pending`, and `second_result.attempt == 2` after the retry run starts and succeeds.
